openflow13/nx_action.go

```diff
@@ -0,0 +1,956 @@
+
+	return err
+}
+
+// nxast_resubmit_table
+type NXActionResubmitTable struct {
+	*NXActionHeader
+	InPort  uint16
+	TableID uint8
+	pad     []byte // 3 bytes
+	withCT  bool
+}
+
+func newNXActionResubmitTable() *NXActionResubmitTable {
+	a := &NXActionResubmitTable{
+		NXActionHeader: NewNxActionHeader(NXAST_RESUBMIT_TABLE),
+		withCT:         false,
+		pad:            make([]byte, 3, 3),
+	}
+	a.Length = 16
+	return a
+}
+
+func newNXActionResubmitTableCT() *NXActionResubmitTable {
+	a := &NXActionResubmitTable{
+		NXActionHeader: NewNxActionHeader(NXAST_RESUBMIT_TABLE),
+		withCT:         true,
+		pad:            make([]byte, 3, 3),
+	}
+	a.Length = 16
+	return a
+}
+
+func NewNXActionResubmitTableAction(inPort uint16, tableId uint8) *NXActionResubmitTable {
+	a := newNXActionResubmitTable()
+	a.InPort = inPort
+	a.TableID = tableId
+	return a
+}
+
+func (a *NXActionResubmitTable) IsCT() bool {
+	return a.withCT
+}
+
+func (a *NXActionResubmitTable) Len() (n uint16) {
+	return a.Length
+}
+
+func (a *NXActionResubmitTable) MarshalBinary() (data []byte, err error) {
+	data = make([]byte, int(a.Len()))
+	b := make([]byte, 0)
+	n := 0
+
+	b, err = a.NXActionHeader.MarshalBinary()
+	copy(data[n:], b)
+	n += len(b)
+	binary.BigEndian.PutUint16(data[n:], a.InPort)
+	n += 2
+	data[n] = a.TableID
+	n += 1
+	copy(data[n:], a.pad)
+	n += 3
+
+	return
+}
+
+func (a *NXActionResubmitTable) UnmarshalBinary(data []byte) error {
+	n := 0
+	a.NXActionHeader = new(NXActionHeader)
+	err := a.NXActionHeader.UnmarshalBinary(data[n:])
+	n += int(a.NXActionHeader.Len())
+	if len(data) < int(a.Len()) {
+		return errors.New("the byte array has wrong size to unmarshal an NXActionResubmitTable message")
+	}
+	a.InPort = binary.BigEndian.Uint16(data[n:])
+	n += 2
+	a.TableID = data[n]
+	n += 1
+	copy(a.pad, data[n:n+3])
+
+	return err
+}
+
+func NewNXActionResubmitTableCT(inPort uint16, tableId uint8) *NXActionResubmitTable {
+	a := newNXActionResubmitTableCT()
+	a.InPort = inPort
+	a.TableID = tableId
+	return a
+}
+
+func NewNXActionResubmitTableCTNoInPort(tableId uint8) *NXActionResubmitTable {
+	a := newNXActionResubmitTableCT()
+	a.InPort = OFPP_IN_PORT
+	a.TableID = tableId
+	return a
+}
+
+// nxast_resubmit_table
+type NXActionCTNAT struct {
+	*NXActionHeader
+	pad          []byte // 2 bytes
+	Flags        uint16 // nat Flags to identify snat/dnat, and nat algorithms: random/protocolHash, connection persistent
+	rangePresent uint16 // mark if has set nat range, including ipv4/ipv6 range, port range
+
+	rangeIPv4Min  net.IP
+	rangeIPv4Max  net.IP
+	rangeIPv6Min  net.IP
+	rangeIPv6Max  net.IP
+	rangeProtoMin *uint16
+	rangeProtoMax *uint16
+}
+
+func NewNXActionCTNAT() *NXActionCTNAT {
+	a := new(NXActionCTNAT)
+	a.NXActionHeader = NewNxActionHeader(NXAST_NAT)
+	a.Length = 16
+	a.pad = make([]byte, 2, 2)
+	return a
+}
+
+func (a *NXActionCTNAT) Len() (n uint16) {
+	a.Length = uint16(math.Ceil(float64(a.Length)/float64(8))) * 8
```

Comment:
I think the following is more concise / elegant:
```
a.Length = ((a.Length + 7) / 8) * 8
```

Reply:
Done.